### ISSUE TITLE
Bad results on custom data

### ISSUE THREAD

**Issue by poornimajd** [nianticlabs/monodepth2#169]
Hello,Great work @mrharicot and team!
I have trained my model on Indian Driving dataset,with rectified stereo images as input, with the following command to train-
 `python3 train.py --model_name finetuned_stereo --load_weights_folder ./stereo_1024x320/ --models_to_load encoder depth --batch_size 12 --use_stereo --frame_ids 0 --split eigen_zhou --log_dir mono2weights/`
I trained for 50 epochs and I got this result
![0000223](https://user-images.githubusercontent.com/31023599/79533769-58b61880-8096-11ea-8eca-c3a5302beb40.jpg)
![0000223_disp](https://user-images.githubusercontent.com/31023599/79533780-5f449000-8096-11ea-8bb3-7d065de92daa.jpeg)
Also I changed the intrinsic parameters and the loss remained same for 5th epoch or 49th epoch as shown 

> Training
> epoch   5 | batch    170 | examples/s:  65.8 | loss: 0.06628 | time elapsed: 03h20m42s | time left: 27h15m49s
> epoch  49 | batch     66 | examples/s:  66.9 | loss: 0.06413 | time elapsed: 29h51m55s | time left: 00h29m51s

Also previously when I had trained with this command:
`python3 train.py --model_name mono_model --batch_size 8 --log_dir /media/smart/'My Passport'/mono2_weights/`
i.e without using --use_stereo label and I used unrectified stereo images as input , I got the following result,with the same input image:
![0000223_disp](https://user-images.githubusercontent.com/31023599/79534282-c1ea5b80-8097-11ea-8c18-21e8386db605.jpeg)
I want to train it using stereo model and stereo rectified images only, But I have not got the expected results, Can someone please suggest what is going wrong?
Any suggestion is greatly appreciated.
Thank you

**Comment by msceme** [comment]
Hi, Can you share the left and right image of some scene?

**Comment by poornimajd** [comment]
These are the left and right images which were given as input.
left
![0000000000l](https://user-images.githubusercontent.com/31023599/79541095-da15a700-80a6-11ea-98d9-7ce4f0d26b27.jpg)
right
![0000000000](https://user-images.githubusercontent.com/31023599/79541105-df72f180-80a6-11ea-9fac-7f96f0db661d.jpg)
I used matlab's uncalibrated rectification code to rectify the images.
looks like this
![mosterror](https://user-images.githubusercontent.com/31023599/79541295-3678c680-80a7-11ea-9c2a-ddc2252019f2.jpg)

**Comment by msceme** [comment]
Looks like objects are moving on both sides. Try replacing sigmoid activation with tanh activation.

**Comment by poornimajd** [comment]
Thanks for the quick response!
The reason you suggested tanh instead of sigmoid is to avoid vanishing gradient and also as the range of tanh is from [-1,1] right?
Also if I change the activation function to tanh in this line:
https://github.com/nianticlabs/monodepth2/blob/13200ab2f29f2f10dec3aa5db29c32a23e29d376/networks/depth_decoder.py#L63
Then the disparity will get computed accordingly but I am a bit confused if disp_to_depth function is to be used to compute depth or is it this formula- depth=f*baseline/disparity . If the function is to be used then the formula will be no longer valid for tanh right?
Any suggestion is greatly appreciated!
Thank you

**Comment by msceme** [comment]
> Thanks for the quick response!
> The reason you suggested tanh instead of sigmoid is to avoid vanishing gradient and also as the range of tanh is from [-1,1] right?
> Also if I change the activation function to tanh in this line:
> https://github.com/nianticlabs/monodepth2/blob/13200ab2f29f2f10dec3aa5db29c32a23e29d376/networks/depth_decoder.py#L63
> 
> 
> Then the disparity will get computed accordingly but I am a bit confused if disp_to_depth function is to be used to compute depth or is it this formula- depth=f*baseline/disparity . If the function is to be used then the formula will be no longer valid for tanh right?
> Any suggestion is greatly appreciated!
> Thank you

Yes. In your case, there are both negative and positive disparities. The formula you mentioned is for parallel camera setup, which only have positive disparities. You have to change the depth formula. The following slides can help you
https://cis.temple.edu/~latecki/Courses/CIS601-03/Lectures/slides_lec9.pdf

**Comment by poornimajd** [comment]
@msceme  thanks for the quick reply and the resource!
I used rectified images as input ,and used tanh as the activation, by changing in this line:

> https://github.com/nianticlabs/monodepth2/blob/13200ab2f29f2f10dec3aa5db29c32a23e29d376/networks/depth_decoder.py#L63

I used the training command:

`python3 train.py --model_name finetuned_stereo --load_weights_folder ./stereo_1024x320/ --models_to_load encoder depth  --min_depth=0.5 --use_stereo --frame_ids 0 --split eigen_zhou --log_dir ./mono2weights/`

I get the following black image as result
![0000223](https://user-images.githubusercontent.com/31023599/79714248-d58eff80-82ed-11ea-919c-4c6928f138dd.jpg)
![0000223_disp](https://user-images.githubusercontent.com/31023599/79714258-da53b380-82ed-11ea-85be-4437722a0227.jpeg)
I am not sure what has gone wrong.
Any suggestion is appreciated.
Thanks

**Comment by msceme** [comment]
> @msceme thanks for the quick reply and the resource!
> I used rectified images as input ,and used tanh as the activation, by changing in this line:
> 
> > https://github.com/nianticlabs/monodepth2/blob/13200ab2f29f2f10dec3aa5db29c32a23e29d376/networks/depth_decoder.py#L63
> 
> I used the training command:
> 
> `python3 train.py --model_name finetuned_stereo --load_weights_folder ./stereo_1024x320/ --models_to_load encoder depth --min_depth=0.5 --use_stereo --frame_ids 0 --split eigen_zhou --log_dir ./mono2weights/`
> 
> I get the following black image as result
> ![0000223](https://user-images.githubusercontent.com/31023599/79714248-d58eff80-82ed-11ea-919c-4c6928f138dd.jpg)
> ![0000223_disp](https://user-images.githubusercontent.com/31023599/79714258-da53b380-82ed-11ea-85be-4437722a0227.jpeg)
> I am not sure what has gone wrong.
> Any suggestion is appreciated.
> Thanks

You have to change the depth function also. Maybe you can use monodepth1.

**Comment by poornimajd** [comment]
Thanks for the quick reply!I am not sure of how to change the depth function,I checked out monodepth1, the disparity computation does not depend on the activation function,so may be I can use that.
Thanks once again!

**Comment by msceme** [comment]
> Thanks for the quick reply!I am not sure of how to change the depth function,I checked out monodepth1, the disparity computation does not depend on the activation function,so may be I can use that.
> Thanks once again!

Use monodepth1 and replace sigmoid with tanh.

**Comment by poornimajd** [comment]
Okay I will check it out.
Thanks for helping out!

**Comment by mrharicot** [comment]
You probably don't need to change the depth function, but only the transform between pairs.
Your stereo images are rectified, rows are aligned within the pair, however infinity doesn't map to zero disparity, which means that you still have some rotation around the Y axis.
Instead of using the stereo mode I would suggest you use the standard monocular training mode.
The network should converge to a single transformation for all the images where there is mostly a translation and a small rotation around the Y axis.
